src/node_io.h

```diff
@@ -0,0 +1,236 @@
+#ifndef SRC_NODE_IO_H_
+#define SRC_NODE_IO_H_
+
+#include "node.h"
+
+namespace node {
+
+namespace io {
+
+enum io_error {
+  // Source would overflow the provided buffer
+  IO_ERROR_OVERFLOW = -256,
+  // Source does not support FLAG_PEEK
+  IO_ERROR_NO_PEEK = -512,
+  // Attempt to pull from an already ended source
+  IO_ERROR_EOF = -768,
+  // Source will not copy data into caller provided buffer
+  IO_ERROR_MUST_NO_COPY = -1024,
+};
+
+enum io_pull_flags {
+  IO_PULL_FLAG_NONE = 0x0,
+  // Callback must be invoked synchronously (also sets the
+  // IO_PULL_FLAG_MUST_COPY and IO_PULL_FLAG_STRICT_LENGTH flags)
+  IO_PULL_FLAG_SYNC = 0xD,
```

Comment:
Why does this imply `IO_PULL_FLAG_MUST_COPY`?

Reply:
Because the only way for the `Source` to provide pointers to it's own buffers is via the callback, which is not being used here. This is telling the `Source`: fill these buffers now and return immediately when you do.

Reply:
So maybe `Pull` could yield its own buffers synchronously, if there are any? This sounds like something that would otherwise create unnecessary memcpy()s otherwise

Reply:
That's actually the intent. The two approaches are:

1. caller allocates the buffer, calls pull, `Source` memcpy's data into those
2. caller calls pull, `Source` provides pointers to it's own buffers without memcpy

*Currently*, for option 2, the `Source` *must* use the callback to deliver it's own buffer pointers.

We could change this, however, so that `Pull` can yield buffers back to allow `Source` to provide it's own pointers without requiring the callback.

Reply:
Btw; one of my (and others’) pain points with streams is that they delay a lot of events, whether necessary or not.

It might be nice if we could drop this requirement, and calls would be allowed to call their callback synchronously or asynchronously, whichever fits more nicely?

Reply:
Yep, I agree. With the callbacks here, that's exactly what I'm allowing for *unless* the code calling `Pull` explicitly says that it needs sync processing. Unless that flag is set, the `Source` is largely free to invoke the callback whenever it wants, sync or async, once the data is available. Unless the code calling `Pull` sets that sync flag, it must not make any assumptions at all about when the callback is going to be invoked.

Reply:
(hopefully I didn't misunderstand your comment there tho :-) ...)